File: ScrollKit/IIMyScene.m
//
//  IIMyScene.m
//  ScrollKit
//
//  Created by Fille Åström on 20/11/13.
//  Copyright (c) 2013 IMGNRY. All rights reserved.
//

#import "IIMyScene.h"

typedef NS_ENUM(NSInteger, IIMySceneZPosition)
{
    kIIMySceneZPositionScrolling = 0,
    kIIMySceneZPositionVerticalAndHorizontalScrolling,
    kIIMySceneZPositionStatic,
};

@interface IIMyScene ()
//kIIMySceneZPositionScrolling
@property (nonatomic, weak) SKSpriteNode *spriteToScroll;
@property (nonatomic, weak) SKSpriteNode *spriteForScrollingGeometry;

//kIIMySceneZPositionStatic
@property (nonatomic, weak) SKSpriteNode *spriteForStaticGeometry;

//kIIMySceneZPositionVerticalAndHorizontalScrolling
@property (nonatomic, weak) SKSpriteNode *spriteToHostHorizontalAndVerticalScrolling;
@property (nonatomic, weak) SKSpriteNode *spriteForHorizontalScrolling;
@property (nonatomic, weak) SKSpriteNode *spriteForVerticalScrolling;
@end

@implementation IIMyScene

-(id)initWithSize:(CGSize)size {
    if (self = [super initWithSize:size]) {

        [self setAnchorPoint:(CGPoint){0,1}];
        SKSpriteNode *spriteToScroll = [SKSpriteNode spriteNodeWithColor:[SKColor clearColor] size:size];
        [spriteToScroll setAnchorPoint:(CGPoint){0,1}];
        [spriteToScroll setZPosition:kIIMySceneZPositionScrolling];
        [self addChild:spriteToScroll];

        //Overlay sprite to make anchor point 0,0 (lower left, default for SK)
        SKSpriteNode *spriteForScrollingGeometry = [SKSpriteNode spriteNodeWithColor:[SKColor clearColor] size:size];
        [spriteForScrollingGeometry setAnchorPoint:(CGPoint){0,0}];
        [spriteForScrollingGeometry setPosition:(CGPoint){0, -size.height}];
        [spriteToScroll addChild:spriteForScrollingGeometry];

        SKSpriteNode *scrollingGeometryAnchorMarker = [SKSpriteNode spriteNodeWithColor:[SKColor redColor] size:(CGSize){20,20}];
        [spriteForScrollingGeometry addChild:scrollingGeometryAnchorMarker];

        SKSpriteNode *spriteForStaticGeometry = [SKSpriteNode spriteNodeWithColor:[SKColor clearColor] size:size];
        [spriteForStaticGeometry setAnchorPoint:(CGPoint){0,0}];
        [spriteForStaticGeometry setPosition:(CGPoint){0, -size.height}];
        [spriteForStaticGeometry setZPosition:kIIMySceneZPositionStatic];
        [self addChild:spriteForStaticGeometry];

        SKSpriteNode *spriteToHostHorizontalAndVerticalScrolling = [SKSpriteNode spriteNodeWithColor:[SKColor clearColor] size:size];
        [spriteToHostHorizontalAndVerticalScrolling setAnchorPoint:(CGPoint){0,0}];
        [spriteToHostHorizontalAndVerticalScrolling setPosition:(CGPoint){0, -size.height}];
        [spriteToHostHorizontalAndVerticalScrolling setZPosition:kIIMySceneZPositionVerticalAndHorizontalScrolling];
        [self addChild:spriteToHostHorizontalAndVerticalScrolling];

        CGSize upAndDownSize = size;
        upAndDownSize.width = 30;
        SKSpriteNode *spriteForVerticalScrolling = [SKSpriteNode spriteNodeWithColor:[SKColor clearColor] size:upAndDownSize];
        [spriteForVerticalScrolling setAnchorPoint:(CGPoint){0,0}];
        [spriteToHostHorizontalAndVerticalScrolling addChild:spriteForVerticalScrolling];

        CGSize leftToRightSize = size;
        leftToRightSize.height = 30;
        SKSpriteNode *spriteForHorizontalScrolling = [SKSpriteNode spriteNodeWithColor:[SKColor clearColor] size:leftToRightSize];
        [spriteForHorizontalScrolling setAnchorPoint:(CGPoint){0,0}];
        [spriteToHostHorizontalAndVerticalScrolling addChild:spriteForHorizontalScrolling];

        //Test sprites for constrained Scrolling
        SKLabelNode *horizontallyScrollingLabel = [SKLabelNode labelNodeWithFontNamed:@"HelveticaNeue"];
        [horizontallyScrollingLabel setText:@"Left and right and not up and down, just left and right because I am so cool and I am an axis label!"];
        [horizontallyScrollingLabel setFontSize:14.0];
        [horizontallyScrollingLabel setFontColor:[SKColor lightGrayColor]];
        [horizontallyScrollingLabel setPosition:(CGPoint){.x = 10.0, .y = 30.0}];
        [horizontallyScrollingLabel setHorizontalAlignmentMode:SKLabelHorizontalAlignmentModeLeft];
        [spriteForHorizontalScrolling addChild:horizontallyScrollingLabel];

        SKLabelNode *verticallyScrollingLabel = [SKLabelNode labelNodeWithFontNamed:@"HelveticaNeue"];
        [verticallyScrollingLabel setText:@"V:50"];
        [verticallyScrollingLabel setFontSize:14.0];
        [verticallyScrollingLabel setFontColor:[SKColor lightGrayColor]];
        [verticallyScrollingLabel setPosition:(CGPoint){.x = 10.0, .y = 50.0}];
        [verticallyScrollingLabel setHorizontalAlignmentMode:SKLabelHorizontalAlignmentModeLeft];
        [verticallyScrollingLabel setZPosition:kIIMySceneZPositionVerticalAndHorizontalScrolling];
        [spriteForVerticalScrolling addChild:verticallyScrollingLabel];

        //Test sprites for scrolling and zooming
        SKSpriteNode *greenTestSprite = [SKSpriteNode spriteNodeWithColor:[SKColor greenColor]
                                                                     size:(CGSize){.width = size.width,
                                                                         .height = size.height*.25}];
        [greenTestSprite setName:@"greenTestSprite"];
        [greenTestSprite setAnchorPoint:(CGPoint){0,0}];
        [spriteForScrollingGeometry addChild:greenTestSprite];

        SKSpriteNode *blueTestSprite = [SKSpriteNode spriteNodeWithColor:[SKColor blueColor]
                                                                    size:(CGSize){.width = size.width*.25,
                                                                        .height = size.height*.25}];
        [blueTestSprite setName:@"blueTestSprite"];
        [blueTestSprite setAnchorPoint:(CGPoint){0,0}];
        [blueTestSprite setPosition:(CGPoint){.x = size.width * .25, .y = size.height *.65}];
        [spriteForScrollingGeometry addChild:blueTestSprite];

        //Test sprites for stationary sprites
        SKLabelNode *stationaryLabel = [SKLabelNode labelNodeWithFontNamed:@"HelveticaNeue"];
        [stationaryLabel setText:@"I'm not gonna move, nope, nope."];
        [stationaryLabel setFontSize:14.0];
        [stationaryLabel setFontColor:[SKColor darkGrayColor]];
        [stationaryLabel setPosition:(CGPoint){.x = 10.0, .y = 60.0}];
        [stationaryLabel setHorizontalAlignmentMode:SKLabelHorizontalAlignmentModeLeft];
        [spriteForStaticGeometry addChild:stationaryLabel];

        //Set properties
        _contentSize = size;
        _spriteToScroll = spriteToScroll;
        _spriteForScrollingGeometry = spriteForScrollingGeometry;
        _spriteForStaticGeometry = spriteForStaticGeometry;
        _spriteToHostHorizontalAndVerticalScrolling = spriteToHostHorizontalAndVerticalScrolling;
        _spriteForVerticalScrolling = spriteForVerticalScrolling;
        _spriteForHorizontalScrolling = spriteForHorizontalScrolling;
        _contentOffset = (CGPoint){0,0};
    }
    return self;
}

-(void)didChangeSize:(CGSize)oldSize
{
    CGSize size = [self size];

    CGPoint lowerLeft = (CGPoint){0, -size.height};

    [self.spriteForStaticGeometry setSize:size];
    [self.spriteForStaticGeometry setPosition:lowerLeft];

    [self.spriteToHostHorizontalAndVerticalScrolling setSize:size];
    [self.spriteToHostHorizontalAndVerticalScrolling setPosition:lowerLeft];
}

-(void)setContentSize:(CGSize)contentSize
{
    if (!CGSizeEqualToSize(contentSize, _contentSize))
    {
        _contentSize = contentSize;
        [self.spriteToScroll setSize:contentSize];
        [self.spriteForScrollingGeometry setSize:contentSize];
        [self.spriteForScrollingGeometry setPosition:(CGPoint){0, -contentSize.height}];

        CGSize verticalSpriteSize = [self.spriteForVerticalScrolling size];
        verticalSpriteSize.height = contentSize.height;
        [self.spriteForVerticalScrolling setSize:verticalSpriteSize];

        CGSize horizontalSpriteSize = [self.spriteForHorizontalScrolling size];
        horizontalSpriteSize.width = contentSize.width;
        [self.spriteForHorizontalScrolling setSize:horizontalSpriteSize];

        [self setContentOffset:self.contentOffset];
    }
}

-(void)setContentOffset:(CGPoint)contentOffset
{
    _contentOffset = contentOffset;
    contentOffset.x *= -1;
    [self.spriteToScroll setPosition:contentOffset];

    CGPoint scrollingLowerLeft = [self.spriteForScrollingGeometry convertPoint:(CGPoint){0,0} toNode:self.spriteToHostHorizontalAndVerticalScrolling];

    CGPoint horizontalScrollingPosition = [self.spriteForHorizontalScrolling position];
    horizontalScrollingPosition.x = scrollingLowerLeft.x;
    [self.spriteForHorizontalScrolling setPosition:horizontalScrollingPosition];

    CGPoint verticalScrollingPosition = [self.spriteForVerticalScrolling position];
    verticalScrollingPosition.y = scrollingLowerLeft.y;
    [self.spriteForVerticalScrolling setPosition:verticalScrollingPosition];
}

-(void)setContentScale:(CGFloat)scale;
{
    [self.spriteToScroll setScale:scale];
}

@end

File: ScrollKit/IIMyScene.swift
//
// ScrollKit
// https://github.com/bobmoff/ScrollKit
// Created by Fille Åström
//

import Foundation
import SpriteKit

enum IIMySceneZPosition: Int {
	case Scrolling = 0
	case VerticalAndHorizontalScrolling
	case Static
}

public class IIMyScene: SKScene {

	var innerContentSize: CGSize = CGSize.zero
	public var contentSize: CGSize {
		get {
			return innerContentSize
		}
		set {
			if !newValue.equalTo(innerContentSize) {
				innerContentSize = newValue
				self.spriteToScroll?.size = newValue
				self.spriteForScrollingGeometry?.size = newValue
				self.spriteForScrollingGeometry?.position = CGPoint.zero
				updateConstrainedScrollerSize()
			}
		}
	}
	
	var innerContentOffset: CGPoint = CGPoint.zero
	public var contentOffset: CGPoint {
		get {
			return innerContentOffset
		}
		set {
			if !newValue.equalTo(innerContentOffset) {
				innerContentOffset = newValue
				contentOffsetReload()
			}
		}
	}
	
	func contentOffsetReload() {
        self.spriteToScroll?.position = CGPoint(x: -innerContentOffset.x, y: -innerContentOffset.y)
		
		if let spriteForScrollingGeometry = self.spriteForScrollingGeometry,
			let spriteToHostHorizontalAndVerticalScrolling = self.spriteToHostHorizontalAndVerticalScrolling,
			let spriteForHorizontalScrolling = self.spriteForHorizontalScrolling,
			let spriteForVerticalScrolling = self.spriteForVerticalScrolling {
				let scrollingLowerLeft = spriteForScrollingGeometry.convert(CGPoint.zero, to: spriteToHostHorizontalAndVerticalScrolling)
				
				var horizontalScrollingPosition = spriteForHorizontalScrolling.position
				horizontalScrollingPosition.y = scrollingLowerLeft.y
				spriteForHorizontalScrolling.position = horizontalScrollingPosition
				
				var verticalScrollingPosition = spriteForVerticalScrolling.position
				verticalScrollingPosition.x = scrollingLowerLeft.x
				spriteForVerticalScrolling.position = verticalScrollingPosition
		}
	}
	
	//kIIMySceneZPositionScrolling
	weak var spriteToScroll: SKSpriteNode?
	weak var spriteForScrollingGeometry: SKSpriteNode?
	
	//kIIMySceneZPositionStatic
	weak var spriteForStaticGeometry: SKSpriteNode?
	
	// ZPosition VerticalAndHorizontalScrolling
	weak var spriteToHostHorizontalAndVerticalScrolling: SKSpriteNode?
	weak var spriteForHorizontalScrolling: SKSpriteNode?
	weak var spriteForVerticalScrolling: SKSpriteNode?

	
	override init(size: CGSize) {
		super.init(size: size)
		self.anchorPoint = CGPoint.zero

		let spriteToScroll = SKSpriteNode(color: SKColor.clear, size: size)
		spriteToScroll.anchorPoint = CGPoint.zero
		spriteToScroll.zPosition = CGFloat(IIMySceneZPosition.Scrolling.rawValue)
		self.addChild(spriteToScroll)
		
		//Overlay sprite to make anchor point 0,0 (lower left, default for SK)
		let spriteForScrollingGeometry = SKSpriteNode(color: SKColor.clear, size: size)
		spriteForScrollingGeometry.anchorPoint = CGPoint.zero
		spriteForScrollingGeometry.position = CGPoint.zero
		spriteForScrollingGeometry.zPosition = CGFloat(IIMySceneZPosition.Scrolling.rawValue)
		spriteToScroll.addChild(spriteForScrollingGeometry)
		
		let spriteForStaticGeometry = SKSpriteNode(color: SKColor.clear, size: size)
		spriteForStaticGeometry.anchorPoint = CGPoint.zero
		spriteForStaticGeometry.position = CGPoint.zero
		spriteForStaticGeometry.zPosition = CGFloat(IIMySceneZPosition.Static.rawValue)
		self.addChild(spriteForStaticGeometry)
		
		let spriteToHostHorizontalAndVerticalScrolling = SKSpriteNode(color: SKColor.clear, size: size)
		spriteToHostHorizontalAndVerticalScrolling.anchorPoint = CGPoint.zero
		spriteToHostHorizontalAndVerticalScrolling.position = CGPoint.zero
		spriteToHostHorizontalAndVerticalScrolling.zPosition = CGFloat(IIMySceneZPosition.VerticalAndHorizontalScrolling.rawValue)
		self.addChild(spriteToHostHorizontalAndVerticalScrolling)

		var upAndDownSize = size
		upAndDownSize.width = 30
		let spriteForVerticalScrolling = SKSpriteNode(color: SKColor.clear, size: upAndDownSize)
		spriteForVerticalScrolling.anchorPoint = CGPoint.zero
        spriteForVerticalScrolling.position = CGPoint(x: 0, y: 30)
		spriteToHostHorizontalAndVerticalScrolling.addChild(spriteForVerticalScrolling)

		var leftToRightSize = size
		leftToRightSize.height = 30
		let spriteForHorizontalScrolling = SKSpriteNode(color: SKColor.clear, size: leftToRightSize)
		spriteForHorizontalScrolling.anchorPoint = CGPoint.zero
        spriteForHorizontalScrolling.position = CGPoint(x: 10, y: 0)
		spriteToHostHorizontalAndVerticalScrolling.addChild(spriteForHorizontalScrolling)
		
		//Test sprites for constrained Scrolling
		var labelPosition = CGFloat(-500.0)
		let stepSize = CGFloat(50.0)
		while labelPosition < 2000.0 {
			let labelText = String(format: "%5.0f", labelPosition)

			let hscrollingLabel = SKLabelNode(fontNamed: "HelveticaNeue")
			hscrollingLabel.text = labelText
			hscrollingLabel.fontSize = 14
			hscrollingLabel.fontColor = SKColor.darkGray
            hscrollingLabel.position = CGPoint(x: 0, y: labelPosition)
			hscrollingLabel.horizontalAlignmentMode = SKLabelHorizontalAlignmentMode.left
			spriteForHorizontalScrolling.addChild(hscrollingLabel)

			let scrollingLabel = SKLabelNode(fontNamed: "HelveticaNeue")
			scrollingLabel.text = labelText
			scrollingLabel.fontSize = 14
			scrollingLabel.fontColor = SKColor.darkGray
            scrollingLabel.position = CGPoint(x: labelPosition, y: 0)
			scrollingLabel.horizontalAlignmentMode = SKLabelHorizontalAlignmentMode.left
			spriteForVerticalScrolling.addChild(scrollingLabel)
			
			labelPosition += stepSize
		}

		//Test sprites for scrolling and zooming
        let greenTestSprite = SKSpriteNode(color: SKColor.green, size: CGSize(width: size.width, height: size.height * 0.25))
		greenTestSprite.name = "greenTestSprite"
		greenTestSprite.anchorPoint = CGPoint.zero
		spriteForScrollingGeometry.addChild(greenTestSprite)
		
        let blueTestSprite = SKSpriteNode(color: SKColor.blue, size: CGSize(width: size.width * 0.25, height: size.height * 0.25))
		blueTestSprite.name = "blueTestSprite"
		blueTestSprite.anchorPoint = CGPoint.zero
        blueTestSprite.position = CGPoint(x: size.width * 0.25, y: size.height * 0.65)
		spriteForScrollingGeometry.addChild(blueTestSprite)
		
		//Test sprites for stationary sprites
		let stationaryLabel = SKLabelNode(fontNamed: "HelveticaNeue")
		stationaryLabel.text = "I'm not gonna move, nope, nope."
		stationaryLabel.fontSize = 14
		stationaryLabel.fontColor = SKColor.darkGray
        stationaryLabel.position = CGPoint(x: 60, y: 80)
		stationaryLabel.horizontalAlignmentMode = SKLabelHorizontalAlignmentMode.left
		spriteForStaticGeometry.addChild(stationaryLabel)
		
		
		//Set properties
		self.contentSize = size
		self.spriteToScroll = spriteToScroll
		self.spriteForScrollingGeometry = spriteForScrollingGeometry
		self.spriteForStaticGeometry = spriteForStaticGeometry
		self.spriteToHostHorizontalAndVerticalScrolling = spriteToHostHorizontalAndVerticalScrolling
		self.spriteForVerticalScrolling = spriteForVerticalScrolling
		self.spriteForHorizontalScrolling = spriteForHorizontalScrolling
		self.contentOffset = CGPoint.zero
	}

	required public init?(coder aDecoder: NSCoder) {
	    fatalError("init(coder:) has not been implemented")
	}
	
	override public func didChangeSize(_ oldSize: CGSize) {
		let size = self.size
		
		
		self.spriteForStaticGeometry?.size = size
		self.spriteForStaticGeometry?.position = CGPoint.zero
		
		self.spriteToHostHorizontalAndVerticalScrolling?.size = size
		self.spriteToHostHorizontalAndVerticalScrolling?.position = CGPoint.zero
	}
	
	func setContentScale(scale: CGFloat) {
		spriteToScroll?.setScale(scale)
		updateConstrainedScrollerSize()
	}
	
	func updateConstrainedScrollerSize() {
		let contentSize: CGSize = self.contentSize
		
		var verticalSpriteSize: CGSize = self.spriteForVerticalScrolling?.size ?? CGSize.zero
		verticalSpriteSize.height = contentSize.height
		spriteForVerticalScrolling?.size = verticalSpriteSize
		
		var horizontalSpriteSize = self.spriteForHorizontalScrolling?.size ?? CGSize.zero
		horizontalSpriteSize.width = contentSize.width
		spriteForHorizontalScrolling?.size = horizontalSpriteSize
		
		let xScale = spriteToScroll?.xScale ?? 1.0
		let yScale = spriteToScroll?.yScale ?? 1.0
		
		spriteForVerticalScrolling?.xScale = xScale
		spriteForVerticalScrolling?.yScale = yScale
		
		spriteForHorizontalScrolling?.xScale = xScale
		spriteForHorizontalScrolling?.yScale = yScale
		
		let xScaleReciprocal = 1.0 / xScale
		let yScaleReciprocal = 1.0 / yScale
		for node in spriteForVerticalScrolling?.children ?? [] {
			node.xScale = xScaleReciprocal
			node.yScale = yScaleReciprocal
		}

		for node in spriteForHorizontalScrolling?.children ?? [] {
			node.xScale = xScaleReciprocal
			node.yScale = yScaleReciprocal
		}

		contentOffsetReload()
	}
}
